xds/src/test/java/io/grpc/xds/LocalityStoreTest.java

```diff
@@ -504,9 +511,112 @@ public PickResult pickSubchannel(PickSubchannelArgs args) {
     assertThat(resolvedAddressesCaptor1.getValue().getAddresses()).containsExactly(eag11);
     assertThat(pickerFactory.totalReadyLocalities).isEqualTo(1);
 
+    fakeClock.forwardTime(14, TimeUnit.MINUTES);
+    verify(loadBalancers.get("sz3"), never()).shutdown();
+    fakeClock.forwardTime(1, TimeUnit.MINUTES);
+    verify(loadBalancers.get("sz3")).shutdown();
+
     verify(random, never()).nextInt(1000_000);
   }
 
+  @Test
+  public void updateLoaclityStore_deactivateAndReactivate() {
+    localityStore.updateDropPercentage(ImmutableList.<DropOverload>of());
+    LocalityInfo localityInfo1 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint11, lbEndpoint12), 1);
+    LocalityInfo localityInfo2 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint21, lbEndpoint22), 2);
+    LocalityInfo localityInfo3 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint31, lbEndpoint32), 3);
+    Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of(
+        locality1, localityInfo1, locality2, localityInfo2, locality3, localityInfo3);
+    localityStore.updateLocalityStore(localityInfoMap);
+
+    assertThat(loadBalancers.keySet()).containsExactly("sz1", "sz2", "sz3");
+
+    final Subchannel subchannel1 = mock(Subchannel.class);
+    SubchannelPicker subchannelPicker1 = new SubchannelPicker() {
+      @Override
+      public PickResult pickSubchannel(PickSubchannelArgs args) {
+        return PickResult.withSubchannel(subchannel1);
+      }
+    };
+    childHelpers.get("sz1").updateBalancingState(READY, subchannelPicker1);
+    final Subchannel subchannel3 = mock(Subchannel.class);
+    SubchannelPicker subchannelPicker3 = new SubchannelPicker() {
+      @Override
+      public PickResult pickSubchannel(PickSubchannelArgs args) {
+        return PickResult.withSubchannel(subchannel3);
+      }
+    };
+    childHelpers.get("sz3").updateBalancingState(READY, subchannelPicker3);
+
+    // update localities, removing sz1, sz2, keeping sz3, and adding sz4
+    LocalityInfo localityInfo4 =
+        new LocalityInfo(ImmutableList.of(lbEndpoint41, lbEndpoint42), 4);
+    localityInfoMap = ImmutableMap.of(locality3, localityInfo3, locality4, localityInfo4);
+    localityStore.updateLocalityStore(localityInfoMap);
+
+    ArgumentCaptor<SubchannelPicker> subchannelPickerCaptor = ArgumentCaptor.forClass(null);
+    // helper updated multiple times. Don't care how many times, just capture the latest picker
+    verify(helper, atLeastOnce()).updateBalancingState(
+        same(READY), subchannelPickerCaptor.capture());
+    assertThat(pickerFactory.totalReadyLocalities).isEqualTo(1);
+    pickerFactory.nextIndex = 0;
+    assertThat(subchannelPickerCaptor.getValue().pickSubchannel(pickSubchannelArgs).getSubchannel())
+        .isEqualTo(subchannel3);
+
+    // verify no traffic will go to deactivated locality
```

Comment:
The test doesn't reflect that those localities are temporarily marked as "deactivated" instead of removed. You may want to verify those "deactivated" child balancers are not shutdown. Otherwise, we cannot tell if the caching mechanism exists.

It is inevitable to test implementation when verifying caching mechanism as itself is an implementation detail.

Reply:
Added verification it's not shutdown.